File: potms-go/render.go
// 模板渲染层 — gonja(Jinja2 兼容引擎) + Flask 风格全局函数
package main

import (
	"fmt"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"github.com/nikolalohinski/gonja/v2"
	"github.com/nikolalohinski/gonja/v2/exec"
	"github.com/nikolalohinski/gonja/v2/loaders"
)

var (
	tmplCache = map[string]*exec.Template{}
	tmplDir   = "templates"
)

// endpointRoutes Flask endpoint → 路径模板（{param} 占位；多余 kwargs 变查询串）
var endpointRoutes = map[string]string{
	"auth.login": "/login", "auth.logout": "/logout", "auth.account": "/account",
	"dashboard.index": "/", "dashboard.backup_now": "/backup/now",
	"personnel.list": "/personnel/", "personnel.info_new": "/personnel/info/new",
	"personnel.info_list":   "/personnel/info/",
	"personnel.info_delete": "/personnel/info/{info_id}/delete",
	"personnel.info_edit":   "/personnel/info/{info_id}/edit",
	"personnel.filing_new":  "/personnel/filing/new",
	"personnel.filing_edit": "/personnel/filing/{filing_id}/edit",
	"personnel.view":        "/personnel/{filing_id}", "personnel.delete": "/personnel/{filing_id}/delete",
	"certificate.list": "/certificate/", "certificate.new": "/certificate/new",
	"certificate.edit": "/certificate/{cert_id}/edit", "certificate.delete": "/certificate/{cert_id}/delete",
	"travel.list": "/travel/", "travel.attachments": "/travel/attachments",
	"travel.new": "/travel/new", "travel.edit": "/travel/{travel_id}/edit",
	"travel.view": "/travel/{travel_id}", "travel.delete": "/travel/{travel_id}/delete",
	"travel.cancel": "/travel/{travel_id}/cancel", "travel.restore": "/travel/{travel_id}/restore",
	"travel.attachment_download": "/travel/attachment/{att_id}/download",
	"travel.attachment_preview":  "/travel/attachment/{att_id}/preview",
	"travel.attachment_delete":   "/travel/attachment/{att_id}/delete",
	"decontrol.list":             "/decontrol/", "decontrol.new": "/decontrol/new/{filing_id}",
	"decontrol.view":     "/decontrol/{dec_id}",
	"export.info_export": "/export/info", "export.filing_export": "/export/filing",
	"export.certificate_export": "/export/certificate", "export.travel_export": "/export/travel",
	"export.decontrol_export": "/export/decontrol",
	"export.print_view":       "/print/{print_type}/{id}", "export.batch_print": "/print/batch/{print_type}",
	"import_data.index": "/import/", "import_data.download_template": "/import/template",
	"logs.index": "/logs/", "logs.export": "/logs/export",
	"organization.index": "/org/", "organization.add": "/org/add",
	"organization.edit": "/org/{org_id}/edit", "organization.delete": "/org/{org_id}/delete",
	"dict_admin.index": "/dict/", "dict_admin.add": "/dict/add",
	"dict_admin.edit": "/dict/{dict_id}/edit", "dict_admin.delete": "/dict/{dict_id}/delete",
	"submit_unit.index": "/submit-unit/", "submit_unit.add": "/submit-unit/add",
	"submit_unit.edit": "/submit-unit/{uid}/edit", "submit_unit.delete": "/submit-unit/{uid}/delete",
	"search.index": "/search",
}

func urlFor(endpoint string, kwargs map[string]string) string {
	if endpoint == "static" {
		return "/static/" + kwargs["filename"]
	}
	path, ok := endpointRoutes[endpoint]
	if !ok {
		return "#"
	}
	var query []string
	for k, v := range kwargs {
		ph := "{" + k + "}"
		if strings.Contains(path, ph) {
			path = strings.ReplaceAll(path, ph, v)
		} else {
			query = append(query, url.QueryEscape(k)+"="+url.QueryEscape(v))
		}
	}
	if len(query) > 0 {
		path += "?" + strings.Join(query, "&")
	}
	return path
}

func loadTemplates() {
	loader, err := loaders.NewFileSystemLoader(tmplDir)
	if err != nil {
		log.Fatal("模板目录加载失败: ", err)
	}
	// 注册 localtime 过滤器（UTC → 本地）
	gonja.DefaultEnvironment.Filters.Register("localtime",
		func(e *exec.Evaluator, in *exec.Value, params *exec.VarArgs) *exec.Value {
			format := "%Y-%m-%d %H:%M:%S"
			if len(params.Args) > 0 {
				format = params.Args[0].String()
			}
			return exec.AsValue(toLocalTime(in.Interface(), format))
		})
	err = filepath.Walk(tmplDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() || !strings.HasSuffix(path, ".html") {
			return err
		}
		rel, _ := filepath.Rel(tmplDir, path)
		rel = filepath.ToSlash(rel)
		tpl, terr := exec.NewTemplate(rel, gonja.DefaultConfig, loader, gonja.DefaultEnvironment)
		if terr != nil {
			return fmt.Errorf("模板 %s 解析失败: %w", rel, terr)
		}
		tmplCache[rel] = tpl
		return nil
	})
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("已加载 %d 个模板", len(tmplCache))
}

// kwStr VarArgs 的 kwargs → map[string]string
func kwStr(params *exec.VarArgs) map[string]string {
	out := map[string]string{}
	for k, v := range params.KwArgs {
		out[k] = v.String()
	}
	return out
}

// baseContext 每请求模板全局（等价 Flask context_processor + Jinja 全局）
func baseContext(w http.ResponseWriter, r *http.Request) map[string]interface{} {
	args := map[string]interface{}{}
	for k, v := range r.URL.Query() {
		if len(v) > 0 {
			args[k] = v[0]
		}
	}
	sess := getSession(r)
	flashes := popFlashes(w, r)

	ctx := map[string]interface{}{
		"session": map[string]interface{}{
			"logged_in": isLoggedIn(r),
			"username":  sess["username"],
		},
		"request": map[string]interface{}{"args": args, "path": r.URL.Path},
		"url_for": func(params *exec.VarArgs) *exec.Value {
			if len(params.Args) == 0 {
				return exec.AsValue("#")
			}
			return exec.AsValue(urlFor(params.Args[0].String(), kwStr(params)))
		},
		"csrf_token": func() string { return csrfToken(w, r) },
		"get_flashed_messages": func(params *exec.VarArgs) *exec.Value {
			var out []interface{}
			for _, f := range flashes {
				out = append(out, []interface{}{f[0], f[1]})
			}
			return exec.AsValue(out)
		},
		// 分页链接：保留当前查询串、替换 page（替代 Jinja 的 **request.args 展开）
		"page_url": func(endpoint string, page int) string {
			q := r.URL.Query()
			q.Set("page", fmt.Sprintf("%d", page))
			return urlFor(endpoint, nil) + "?" + q.Encode()
		},
		"nav_q": func() string {
			if r.URL.Path == "/search" {
				return r.URL.Query().Get("q")
			}
			return ""
		}(),
		"dict_opts":      func(cat string) interface{} { return rowsIface(getDictOptions(cat)) },
		"dict_value":     getDictValue,
		"org_flat":       func() interface{} { return rowsIface(getOrgFlat()) },
		"org_tree_opts":  func() interface{} { return rowsIface(getOrgTreeOptions()) },
		"org_children":   func(pid int64) interface{} { return rowsIface(getOrgChildren(pid)) },
		"personnel_opts": func() interface{} { return rowsIface(getPersonnelOptions()) },
		"submit_units":   func() interface{} { return rowsIface(getSubmitUnits()) },
	}
	return ctx
}

func rowsIface(rows []Row) []interface{} {
	out := make([]interface{}, len(rows))
	for i, r := range rows {
		out[i] = r
	}
	return out
}

// render 渲染模板并写响应
func render(w http.ResponseWriter, r *http.Request, name string, data Row) {
	renderStatus(w, r, name, data, http.StatusOK)
}

func renderStatus(w http.ResponseWriter, r *http.Request, name string, data Row, status int) {
	tpl, ok := tmplCache[name]
	if !ok {
		http.Error(w, "模板不存在: "+name, 500)
		return
	}
	ctx := exec.NewContext(baseContext(w, r))
	if data != nil {
		ctx.Update(exec.NewContext(map[string]interface{}(data)))
	}
	out, err := tpl.ExecuteToString(ctx)
	if err != nil {
		log.Printf("模板渲染失败 %s: %v", name, err)
		serverError(w, r)
		return
	}
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.WriteHeader(status)
	w.Write([]byte(out))
}

func notFound(w http.ResponseWriter, r *http.Request) {
	if tpl, ok := tmplCache["errors/404.html"]; ok {
		ctx := exec.NewContext(baseContext(w, r))
		if out, err := tpl.ExecuteToString(ctx); err == nil {
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			w.WriteHeader(404)
			w.Write([]byte(out))
			return
		}
	}
	http.NotFound(w, r)
}

func serverError(w http.ResponseWriter, r *http.Request) {
	if tpl, ok := tmplCache["errors/500.html"]; ok {
		ctx := exec.NewContext(baseContext(w, r))
		if out, err := tpl.ExecuteToString(ctx); err == nil {
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			w.WriteHeader(500)
			w.Write([]byte(out))
			return
		}
	}
	http.Error(w, "Internal Server Error", 500)
}

func redirect(w http.ResponseWriter, r *http.Request, endpoint string, kwargs map[string]string) {
	http.Redirect(w, r, urlFor(endpoint, kwargs), http.StatusFound)
}

File: potms-rust/src/render.rs
// 渲染层 — minijinja(Jinja2 兼容) + Flask 风格全局函数
use crate::config::Config;
use crate::helpers;
use axum::response::{Html, IntoResponse, Response};
use minijinja::value::{Kwargs, Rest};
use minijinja::{Environment, State, Value};
use rusqlite::Connection;
use serde_json::json;
use std::sync::{Arc, Mutex};

pub type Db = Arc<Mutex<Connection>>;

#[derive(rust_embed::RustEmbed)]
#[folder = "templates/"]
pub struct Templates;

#[derive(rust_embed::RustEmbed)]
#[folder = "static/"]
pub struct StaticAssets;

// Flask endpoint → 路径模板（{param} 占位；多余 kwargs 变查询串）
pub fn endpoint_path(endpoint: &str) -> Option<&'static str> {
    Some(match endpoint {
        "auth.login" => "/login", "auth.logout" => "/logout", "auth.account" => "/account",
        "dashboard.index" => "/", "dashboard.backup_now" => "/backup/now",
        "personnel.list" => "/personnel/", "personnel.info_new" => "/personnel/info/new",
        "personnel.info_list" => "/personnel/info/",
        "personnel.info_delete" => "/personnel/info/{info_id}/delete",
        "personnel.info_edit" => "/personnel/info/{info_id}/edit",
        "personnel.filing_new" => "/personnel/filing/new",
        "personnel.filing_edit" => "/personnel/filing/{filing_id}/edit",
        "personnel.view" => "/personnel/{filing_id}",
        "personnel.delete" => "/personnel/{filing_id}/delete",
        "certificate.list" => "/certificate/", "certificate.new" => "/certificate/new",
        "certificate.edit" => "/certificate/{cert_id}/edit",
        "certificate.delete" => "/certificate/{cert_id}/delete",
        "travel.list" => "/travel/", "travel.attachments" => "/travel/attachments",
        "travel.new" => "/travel/new", "travel.edit" => "/travel/{travel_id}/edit",
        "travel.view" => "/travel/{travel_id}", "travel.delete" => "/travel/{travel_id}/delete",
        "travel.cancel" => "/travel/{travel_id}/cancel", "travel.restore" => "/travel/{travel_id}/restore",
        "travel.attachment_download" => "/travel/attachment/{att_id}/download",
        "travel.attachment_preview" => "/travel/attachment/{att_id}/preview",
        "travel.attachment_delete" => "/travel/attachment/{att_id}/delete",
        "decontrol.list" => "/decontrol/", "decontrol.new" => "/decontrol/new/{filing_id}",
        "decontrol.view" => "/decontrol/{dec_id}",
        "export.info_export" => "/export/info", "export.filing_export" => "/export/filing",
        "export.certificate_export" => "/export/certificate", "export.travel_export" => "/export/travel",
        "export.decontrol_export" => "/export/decontrol",
        "export.print_view" => "/print/{print_type}/{id}", "export.batch_print" => "/print/batch/{print_type}",
        "import_data.index" => "/import/", "import_data.download_template" => "/import/template",
        "logs.index" => "/logs/", "logs.export" => "/logs/export",
        "organization.index" => "/org/", "organization.add" => "/org/add",
        "organization.edit" => "/org/{org_id}/edit", "organization.delete" => "/org/{org_id}/delete",
        "dict_admin.index" => "/dict/", "dict_admin.add" => "/dict/add",
        "dict_admin.edit" => "/dict/{dict_id}/edit", "dict_admin.delete" => "/dict/{dict_id}/delete",
        "submit_unit.index" => "/submit-unit/", "submit_unit.add" => "/submit-unit/add",
        "submit_unit.edit" => "/submit-unit/{uid}/edit", "submit_unit.delete" => "/submit-unit/{uid}/delete",
        "search.index" => "/search",
        _ => return None,
    })
}

pub fn url_for(endpoint: &str, kwargs: &[(String, String)]) -> String {
    if endpoint == "static" {
        let filename = kwargs.iter().find(|(k, _)| k == "filename").map(|(_, v)| v.clone()).unwrap_or_default();
        return format!("/static/{filename}");
    }
    let mut path = match endpoint_path(endpoint) {
        Some(p) => p.to_string(),
        None => return "#".into(),
    };
    let mut query: Vec<String> = vec![];
    for (k, v) in kwargs {
        let ph = format!("{{{k}}}");
        if path.contains(&ph) {
            path = path.replace(&ph, v);
        } else {
            query.push(format!("{}={}", urlencoding::encode(k), urlencoding::encode(v)));
        }
    }
    if !query.is_empty() {
        path.push('?');
        path.push_str(&query.join("&"));
    }
    path
}

pub fn build_env(db: Db, cfg: Config) -> Environment<'static> {
    let mut env = Environment::new();
    env.set_auto_escape_callback(|name| {
        if name.ends_with(".html") {
            minijinja::AutoEscape::Html
        } else {
            minijinja::AutoEscape::None
        }
    });
    for name in Templates::iter() {
        let file = Templates::get(&name).unwrap();
        let body = String::from_utf8_lossy(&file.data).into_owned();
        env.add_template_owned(name.to_string(), body).expect("模板解析失败");
    }

    // localtime 过滤器
    let off = cfg.tz_offset_hours;
    env.add_filter("localtime", move |v: Value, args: Rest<Value>| -> Value {
        let s = v.as_str().map(|x| x.to_string()).unwrap_or_else(|| v.to_string());
        let fmt = args.first().and_then(|x| x.as_str()).unwrap_or("%Y-%m-%d %H:%M:%S");
        Value::from(helpers::to_local_time(&s, fmt, off))
    });

    // url_for
    env.add_function("url_for", |endpoint: &str, kwargs: Kwargs| -> Result<Value, minijinja::Error> {
        let mut pairs = vec![];
        for k in kwargs.args() {
            let val: Value = kwargs.get(k)?;
            let sv = val.as_str().map(|x| x.to_string()).unwrap_or_else(|| val.to_string());
            pairs.push((k.to_string(), sv));
        }
        Ok(Value::from(url_for(endpoint, &pairs)))
    });

    // csrf_token()：读取 State 中的 _csrf
    env.add_function("csrf_token", |state: &State| -> Value {
        state.lookup("_csrf").unwrap_or_else(|| Value::from(""))
    });

    // get_flashed_messages()：读取 State 中的 _flashes
    env.add_function("get_flashed_messages", |state: &State, _kwargs: Kwargs| -> Value {
        state.lookup("_flashes").unwrap_or_else(|| Value::from(Vec::<Value>::new()))
    });

    // page_url(endpoint, page)：保留当前查询串、替换 page
    env.add_function("page_url", |state: &State, endpoint: &str, page: i64| -> Value {
        let q = state.lookup("_query").and_then(|v| v.as_str().map(|s| s.to_string())).unwrap_or_default();
        let mut params: Vec<(String, String)> = vec![];
        for pair in q.split('&').filter(|s| !s.is_empty()) {
            if let Some((k, v)) = pair.split_once('=') {
                if k != "page" {
                    params.push((k.to_string(), v.to_string()));
                }
            }
        }
        let base = url_for(endpoint, &[]);
        let mut qs = params.iter().map(|(k, v)| format!("{k}={v}")).collect::<Vec<_>>();
        qs.push(format!("page={page}"));
        Value::from(format!("{base}?{}", qs.join("&")))
    });

    // DB 数据源全局（捕获 db）
    let d = db.clone();
    env.add_function("dict_opts", move |cat: &str| -> Value {
        let conn = d.lock().unwrap();
        to_val(helpers::get_dict_options(&conn, cat))
    });
    let d = db.clone();
    env.add_function("dict_value", move |cat: &str, code: &str| -> Value {
        let conn = d.lock().unwrap();
        Value::from(helpers::get_dict_value(&conn, cat, code))
    });
    let d = db.clone();
    env.add_function("org_flat", move || -> Value {
        let conn = d.lock().unwrap();
        to_val(helpers::get_org_flat(&conn))
    });
    let d = db.clone();
    env.add_function("org_tree_opts", move || -> Value {
        let conn = d.lock().unwrap();
        to_val(helpers::get_org_tree_options(&conn))
    });
    let d = db.clone();
    env.add_function("org_children", move |pid: i64| -> Value {
        let conn = d.lock().unwrap();
        to_val(crate::db::query_maps(&conn, "SELECT id, name FROM sys_org WHERE parent_id = ? ORDER BY sort_order", &[rusqlite::types::Value::Integer(pid)]))
    });
    let d = db.clone();
    env.add_function("personnel_opts", move || -> Value {
        let conn = d.lock().unwrap();
        to_val(helpers::get_personnel_options(&conn))
    });
    let d = db.clone();
    env.add_function("submit_units", move || -> Value {
        let conn = d.lock().unwrap();
        to_val(helpers::get_submit_units(&conn))
    });

    env
}

fn to_val(rows: Vec<serde_json::Value>) -> Value {
    Value::from_serialize(&rows)
}

// 每请求基础上下文（session/request/flashes/csrf/query），与 handler 数据合并
pub struct Ctx {
    pub base: serde_json::Map<String, serde_json::Value>,
}

pub fn base_context(sess: &mut crate::session::Session, path: &str, query: &str) -> Ctx {
    let mut args = serde_json::Map::new();
    for pair in query.split('&').filter(|s| !s.is_empty()) {
        if let Some((k, v)) = pair.split_once('=') {
            args.insert(k.to_string(), json!(urlencoding::decode(v).unwrap_or_default().into_owned()));
        }
    }
    let csrf = sess.csrf_token();
    let flashes: Vec<serde_json::Value> = sess
        .pop_flashes()
        .into_iter()
        .map(|(c, m)| json!([c, m]))
        .collect();
    let nav_q = if path == "/search" {
        args.get("q").and_then(|v| v.as_str()).unwrap_or("").to_string()
    } else {
        String::new()
    };
    let mut base = serde_json::Map::new();
    base.insert("session".into(), json!({"logged_in": sess.logged_in(), "username": sess.username()}));
    base.insert("request".into(), json!({"args": args, "path": path, "endpoint": ""}));
    base.insert("_csrf".into(), json!(csrf));
    base.insert("_flashes".into(), json!(flashes));
    base.insert("_query".into(), json!(query));
    base.insert("nav_q".into(), json!(nav_q));
    Ctx { base }
}

pub fn render(env: &Environment, mut ctx: Ctx, sess: &crate::session::Session, cfg: &Config, name: &str, data: serde_json::Value) -> Response {
    render_status(env, &mut ctx, sess, cfg, name, data, axum::http::StatusCode::OK)
}

pub fn render_status(env: &Environment, ctx: &mut Ctx, sess: &crate::session::Session, cfg: &Config, name: &str, data: serde_json::Value, status: axum::http::StatusCode) -> Response {
    if let serde_json::Value::Object(m) = data {
        for (k, v) in m {
            ctx.base.insert(k, v);
        }
    }
    let tmpl = match env.get_template(name) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("模板不存在 {name}: {e}");
            return (axum::http::StatusCode::INTERNAL_SERVER_ERROR, "模板不存在").into_response();
        }
    };
    let out = match tmpl.render(serde_json::Value::Object(ctx.base.clone())) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("模板渲染失败 {name}: {e:#}");
            return (axum::http::StatusCode::INTERNAL_SERVER_ERROR, "Internal Server Error").into_response();
        }
    };
    let mut resp = (status, Html(out)).into_response();
    attach_session(&mut resp, sess, cfg);
    resp
}

pub fn attach_session(resp: &mut Response, sess: &crate::session::Session, cfg: &Config) {
    if sess.dirty {
        if let Ok(hv) = axum::http::HeaderValue::from_str(&sess.to_cookie(&cfg.secret_key)) {
            resp.headers_mut().append(axum::http::header::SET_COOKIE, hv);
        }
    }
}

pub fn redirect(sess: &crate::session::Session, cfg: &Config, endpoint: &str, kwargs: &[(String, String)]) -> Response {
    let loc = url_for(endpoint, kwargs);
    let mut resp = axum::response::Redirect::to(&loc).into_response();
    attach_session(&mut resp, sess, cfg);
    resp
}
